### PR TITLE
Rebase state prefixdb

### PR DESCRIPTION
**Pull request by aaronbuchwald**

### CODE REVIEW

File: vms/avm/state.go [hunk 1]
```diff
@@ -15,29 +16,24 @@ var (
...
-	cacher.Put(id, uID)
-	return uID
-}
-
 // state is a thin wrapper around a database to provide, caching, serialization,
 // and de-serialization.
-type state struct{ avax.State }
+type state struct {
+	txCache cache.Cacher
+	txDb    database.Database
```
Comment (StephenButtolph) [inline vms/avm/state.go:23]: `txDb` -> `txDB`

File: vms/avm/vm.go [hunk 2]
```diff
@@ -190,17 +199,20 @@ func (vm *VM) Initialize(
...
+		state: &state{
+			txCache: &cache.LRU{Size: txCacheSize},
+			txDb:    prefixdb.NewNested([]byte{txDbPrefix}, vm.db),
+			State: avax.State{
+				UTXOCache:    &cache.LRU{Size: utxoCacheSize},
+				UTXODB:       prefixdb.NewNested([]byte{utxoDbPrefix}, vm.db),
+				StatusCache:  &cache.LRU{Size: statusCacheSize},
+				StatusDB:     prefixdb.NewNested([]byte{statusDbPrefix}, vm.db),
+				IDCache:      &cache.LRU{Size: idCacheSize},
+				IDDB:         prefixdb.NewNested([]byte{idDbPrefix}, vm.db),
```
Comment (StephenButtolph) [inline vms/avm/vm.go:211]: Is there a reason why `NewNested` is used here?

File: vms/avm/vm.go [hunk 3]
```diff
@@ -64,6 +66,13 @@ var (
 	_ vertex.DAGVM = &VM{}
 )
 
+const (
+	utxoDbPrefix byte = iota
+	txDbPrefix
+	statusDbPrefix
+	idDbPrefix
+)
```
Comment (StephenButtolph) [inline vms/avm/vm.go:74]: It seems like in other places strings are used. Let's try to keep it them same everywhere.

Comment (danlaine) [reply vms/avm/vm.go:74]: This avoids unnecessary byte copying but I'm alright with using strings, too

Comment (aaronbuchwald) [reply vms/avm/vm.go:74]: I'll revert this in that case.

File: vms/components/avax/prefixed_state.go [hunk 4]
```diff
@@ -296,32 +114,32 @@ func (s *State) Status(id ids.ID) (choices.Status, error) {
...
 }
 
-// IDs returns the slice of IDs associated with [id], starting after [start].
+// IDs returns the slice of IDs associated with [key], starting after [start].
 // If start is ids.Empty, starts at beginning.
 // Returns at most [limit] IDs.
 func (s *State) IDs(key []byte, start []byte, limit int) ([]ids.ID, error) {
 	idSlice := []ids.ID(nil)
-	iter := prefixdb.NewNested(key, s.DB).NewIteratorWithStart(start)
+	iter := prefixdb.NewNested(key, s.IDDB).NewIteratorWithStart(start)
```
Comment (StephenButtolph) [inline vms/components/avax/prefixed_state.go:142]: Can we properly close this DB here?

File: vms/components/avax/prefixed_state.go [hunk 5]
```diff
@@ -338,12 +156,10 @@ func (s *State) IDs(key []byte, start []byte, limit int) ([]ids.ID, error) {
 
 // AddID saves an ID to the prefixed database
 func (s *State) AddID(key []byte, id ids.ID) error {
-	db := prefixdb.NewNested(key, s.DB)
-	return db.Put(id[:], nil)
+	return prefixdb.NewNested(key, s.IDDB).Put(id[:], nil)
```
Comment (StephenButtolph) [inline vms/components/avax/prefixed_state.go:159]: Can we properly close this DB here?

File: vms/components/avax/prefixed_state.go [hunk 6]
```diff
@@ -338,12 +156,10 @@ func (s *State) IDs(key []byte, start []byte, limit int) ([]ids.ID, error) {
...
 }
 
 // RemoveID removes an ID from the prefixed database
 func (s *State) RemoveID(key []byte, id ids.ID) error {
-	db := prefixdb.NewNested(key, s.DB)
-	return db.Delete(id[:])
+	return prefixdb.NewNested(key, s.IDDB).Delete(id[:])
```
Comment (StephenButtolph) [inline vms/components/avax/prefixed_state.go:164]: Can we properly close this DB here?

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**